File: apps/ios/ExpCore/Sources/API/TeamInvitesApi.swift
import Foundation

// Invite LINKS, both directions (EXP-725): the app mints one (`create` →
// the raw token, wrapped into `<instance>/invite/<token>` by
// `WebLinks.invite`) and accepts one. Link-only on purpose — emailed invites
// stay a web surface, and the mint is shown ONLY while the team has free
// seats: at the cap the control is REMOVED rather than explained, so no
// seat/billing copy can reach an App Store build (3.1.1). The server is still
// the authority: a mint over the cap throws PRECONDITION_FAILED
// (`Error.isPlanLimitError`).

public struct AcceptInviteInput: Encodable, Sendable {
    public let token: String

    public init(token: String) {
        self.token = token
    }
}

public struct CreateInviteInput: Encodable, Sendable {
    public let teamId: String
    /// The server defaults to `member`; sent explicitly because the app never
    /// offers the owner role here.
    public let role: String

    public init(teamId: String, role: String = "member") {
        self.teamId = teamId
        self.role = role
    }
}

/// Only the token is decoded — the invite row itself arrives over the
/// `team_invites` shape, and the token is a bearer secret the shape's column
/// allowlist deliberately drops.
public struct CreateInviteResult: Decodable, Sendable {
    public let token: String

    public init(token: String) {
        self.token = token
    }
}

public final class TeamInvitesApi: Sendable {
    private let trpc: TrpcClient

    public init(trpc: TrpcClient) {
        self.trpc = trpc
    }

    public func accept(accountId: String, token: String) async throws {
        try await trpc.mutationVoid(accountId: accountId, path: "teamInvites.accept", input: AcceptInviteInput(token: token))
    }

    /// Mint a member invite link (owner-only server-side). Returns the raw
    /// token; `WebLinks.invite` turns it into the shareable URL.
    public func create(accountId: String, teamId: String) async throws -> String {
        let result: CreateInviteResult = try await trpc.mutation(
            accountId: accountId,
            path: "teamInvites.create",
            input: CreateInviteInput(teamId: teamId)
        )
        return result.token
    }
}

File: apps/android/app/src/main/java/com/exponential/app/data/api/TeamInvitesApi.kt
package com.exponential.app.data.api

import com.exponential.app.data.db.TeamEntity
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.serialization.Serializable

// Invite links: mint (EXP-725), preview and accept. Creating one is owner-only
// server-side and the app only OFFERS it while the team has free seats — at
// the cap the control is removed entirely, so a store build never renders the
// seat-cap billing copy (EXP-216).

@Serializable
data class AcceptInviteInput(val token: String)

@Serializable
data class AcceptInviteResult(
    val team: TeamEntity,
    val alreadyMember: Boolean = false,
)

@Serializable
data class GetByTokenInput(val token: String)

@Serializable
data class InvitePreview(
    val id: String,
    val teamId: String,
    val role: String,
    val acceptedAt: String? = null,
    val expiresAt: String,
    val teamName: String,
)

@Serializable
data class GetByTokenResult(val invite: InvitePreview)

// `role` always rides explicitly rather than leaning on the server default —
// the wire shape is the contract, and a member invite is what every client
// mints. The result also carries the invite row and `emailDelivered`; only the
// token is decoded here (the app never types an address into the invite).
@Serializable
data class CreateInviteInput(val teamId: String, val role: String = "member")

@Serializable
data class CreateInviteResult(val token: String)

@Singleton
class TeamInvitesApi @Inject constructor(private val trpc: TrpcClient) {

    /** Mint a shareable invite link token (owner-only server-side). */
    suspend fun create(accountId: String, teamId: String): String =
        trpc.mutation(
            accountId,
            path = "teamInvites.create",
            input = CreateInviteInput(teamId),
            inputSerializer = CreateInviteInput.serializer(),
            outputSerializer = CreateInviteResult.serializer(),
        ).token

    suspend fun accept(accountId: String, token: String): AcceptInviteResult =
        trpc.mutation(
            accountId,
            path = "teamInvites.accept",
            input = AcceptInviteInput(token),
            inputSerializer = AcceptInviteInput.serializer(),
            outputSerializer = AcceptInviteResult.serializer(),
        )

    suspend fun getByToken(accountId: String, token: String): InvitePreview =
        trpc.query(
            accountId,
            path = "teamInvites.getByToken",
            input = GetByTokenInput(token),
            inputSerializer = GetByTokenInput.serializer(),
            outputSerializer = GetByTokenResult.serializer(),
        ).invite
}
